src/desktop.cpp

```diff
@@ -65,45 +65,257 @@ dxp_desktop::save_screen ()
+                        int source_height,
+                        int target_width, /* Target dimensions */
+                        int target_height)
 {
-  // TODO(mmskv): not sure what this variable is meant to represent
-  constexpr size_t k_half_int = 16;
+  const auto *input32 = reinterpret_cast<const uint32_t *> (input);
+  auto *output32 = reinterpret_cast<uint32_t *> (output);
 
-  const int x_ratio = (source_width << k_half_int) / target_width;
-  const int y_ratio = (source_height << k_half_int) / target_height;
-  const int colors = 4;
+  //
+  // Bitshifts are used to preserve precision in x_ratio and y_ratio.
+  // Straight up dividing source_width by target_width will lose some data.
+  // Bitshifting this value left increases precision after the division.
+  // And when the real x_ratio needs to be used it can be bitshifted right.
+  //
+  // k_precision_bytes is the number of bits to reserve for precision.
+  // Can be way lower than 16.
+  //
+  constexpr int k_precision_bytes = 16;
+
+  const int x_ratio = (source_width << k_precision_bytes) / target_width;
+  const int y_ratio = (source_height << k_precision_bytes) / target_height;
 
   for (int y = 0; y < target_height; y++)
     {
-      int y2_xsource = ((y * y_ratio) >> k_half_int) * source_width;
-      int i_xdest = y * target_width;
+      int y_source = ((y * y_ratio) >> k_precision_bytes) * source_width;
+      int y_dest = y * target_width;
 
+      int x_source = 0;
+      const uint32_t *input32_line = input32 + y_source;
       for (int x = 0; x < target_width; x++)
         {
-          int x2 = ((x * x_ratio) >> k_half_int);
-          int y2_x2_colors = (y2_xsource + x2) * colors;
-          int i_x_colors = (i_xdest + x) * colors;
-
-          output[i_x_colors] = input[y2_x2_colors];
-          output[i_x_colors + 1] = input[y2_x2_colors + 1];
-          output[i_x_colors + 2] = input[y2_x2_colors + 2];
-          output[i_x_colors + 3] = input[y2_x2_colors + 3];
+          x_source += x_ratio;
+          output32[y_dest + x] = input32_line[x_source >> k_precision_bytes];
+        }
+    }
+}
+
+/**
+ * Bilinear resize.
+ * Additional overhead compared to nearest neighbor resize does not result in
+ * significantly better images. So this resize algorithm is not used.
+ */
+void
+dxp_desktop::bilinear_resize (const uint8_t *__restrict input,
+                              uint8_t *__restrict output,
+                              int source_width, /* Source dimensions */
+                              int source_height,
+                              int target_width, /* Target dimensions */
+                              int target_height)
+{
+  const float x_ratio = float (source_width) / target_width;
+  const float y_ratio = float (source_height) / target_height;
+
+  for (int y_dst = 0; y_dst < target_height; y_dst++)
+    {
+      float y_src = y_dst * y_ratio;
+      int y = std::floor (y_src);
+      float dy = y_src - y;
```

Comment:
are you sure it is the best way to name it?

Reply:
Any suggestions?